compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchTensorPass.cpp

```diff
@@ -0,0 +1,221 @@
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/Transforms/Transforms.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-flow-tensor-to-flow"
+
+namespace mlir::iree_compiler::IREE::Flow {
+
+#define GEN_PASS_DEF_DISPATCHTENSORPASS
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h.inc"
+
+/// Return `true` if the given op is contained in DispatchWorkgroupsOp or in a
+/// DispatchRegionOp.
+static bool isInDispatchRegion(Operation *op) {
+  return op->getParentOfType<IREE::Flow::DispatchWorkgroupsOp>() ||
+         op->getParentOfType<IREE::Flow::DispatchRegionOp>();
+}
+
+/// Wrap a single op in a DispatchWorkgroupsOp.
+static FailureOr<IREE::Flow::DispatchWorkgroupsOp>
+wrapInWorkgroupsOp(mlir::TensorDimTrackingRewriter &rewriter, Operation *op) {
+  // Simplify tensor::DimOps.
+  SmallVector<tensor::DimOp> dimOps = rewriter.getTensorDimOps();
+  if (failed(iree_compiler::IREE::Flow::simplifyDimOps(
+          rewriter, rewriter.getTensorDimOps())))
+    return failure();
+
+  // Wrap operation.
+  auto regionOp = IREE::Flow::wrapOpInDispatchRegion(rewriter, op);
+  if (failed(regionOp))
+    return failure();
+  if (failed(cloneProducersToRegion(rewriter, *regionOp)))
+    return failure();
+  auto workgroupsOp =
+      IREE::Flow::rewriteFlowDispatchRegionToFlowDispatchWorkgroups(*regionOp,
+                                                                    rewriter);
+  if (failed(workgroupsOp))
+    return failure();
+  return *workgroupsOp;
+}
+
+/// Wrap all given ops in a DispatchWorkgroupsOp.
+static FailureOr<SmallVector<IREE::Flow::DispatchWorkgroupsOp>>
+wrapInWorkgroupsOp(mlir::TensorDimTrackingRewriter &rewriter,
+                   SmallVector<Operation *> rootOps) {
+  SmallVector<IREE::Flow::DispatchWorkgroupsOp> result;
+  for (Operation *rootOp : rootOps) {
+    auto workgroupsOp = wrapInWorkgroupsOp(rewriter, rootOp);
+    if (failed(workgroupsOp))
+      return failure();
+    result.push_back(*workgroupsOp);
+  }
+  return result;
+}
+
+/// Wrap all ops of the given types that are direct children of the given op
+/// in DispatchWorkgroupsOps.
+template <typename... OpTys>
+static FailureOr<SmallVector<IREE::Flow::DispatchWorkgroupsOp>>
+wrapInWorkgroupsOp(mlir::TensorDimTrackingRewriter &rewriter, Operation *op) {
+  // Find ops of type OpTys.
+  SmallVector<Operation *> rootOps;
+  for (Region &r : op->getRegions())
+    for (Block &b : r.getBlocks())
+      for (Operation &op : b)
+        if (isa<OpTys...>(&op))
+          rootOps.push_back(&op);
```

Comment:
Can we use the `getOps<>()` method from the region? Who is using this method?

Reply:
Yeah, your right its not being used. Just deleting it